tensorflow/core/kernels/fixed_length_record_reader_op.cc

```diff
@@ -38,55 +42,178 @@ class FixedLengthRecordReader : public ReaderBase {
         hop_bytes_(hop_bytes),
         env_(env),
         file_pos_limit_(-1),
-        record_number_(0) {}
+        record_number_(0),
+        encoding_(encoding) {}
 
   // On success:
-  // * input_buffer_ != nullptr,
-  // * input_buffer_->Tell() == header_bytes_
+  // * buffered_input_stream_ != nullptr,
+  // * buffered_input_stream_->Tell() == header_bytes_
   // * file_pos_limit_ == file size - footer_bytes_
   Status OnWorkStartedLocked() override {
     record_number_ = 0;
-    uint64 file_size = 0;
-    TF_RETURN_IF_ERROR(env_->GetFileSize(current_work(), &file_size));
-    file_pos_limit_ = file_size - footer_bytes_;
+    if (encoding_ == "ZLIB" || encoding_ == "GZIP") {
+      const io::ZlibCompressionOptions zlib_options =
+          encoding_ == "ZLIB" ?
+          io::ZlibCompressionOptions::DEFAULT() :
+          io::ZlibCompressionOptions::GZIP();
 
-    TF_RETURN_IF_ERROR(env_->NewRandomAccessFile(current_work(), &file_));
+      file_pos_limit_ = -1;
+      hop_cache_.clear();
+      footer_cache_.clear();
+
+      TF_RETURN_IF_ERROR(env_->NewRandomAccessFile(current_work(), &file_));
+      file_stream_.reset(new io::RandomAccessInputStream(file_.get()));
+      buffered_input_stream_.reset(new io::ZlibInputStream(file_stream_.get(),
+                                                          (size_t)kBufferSize,
+                                                          (size_t)kBufferSize,
+                                                          zlib_options));
+      TF_RETURN_IF_ERROR(buffered_input_stream_->SkipNBytes(header_bytes_));
+      // The following is similiar to non encoding case except that we also caches
+      // footer.
+      // Basically, each read only advances hop_bytes_. However, we have the data
+      // that is pieced together by
+      // hop_cache_ + footer_cache_ + read(hop_bytes_)
+      // Then we use the following to cut data for use:
+      // value:                     0 --> record_bytes_
+      // hop_cache_:       hop_bytes_ --> record_bytes_
+      // footer_cache_: record_bytes_ --> end(record_bytes_ + footer_bytes_)
+      // Next time we read, we advance hop_bytes_ and append to the end, and
+      // repeat the same process.
+      if (0 < hop_bytes_ && hop_bytes_ < record_bytes_) {
```

Comment:
Thanks for intricately handling hop_bytes and footer_bytes. It may be cleaner to combine the two caches into a `lookahead_cache_` which could be of size `record_bytes_ + footer_bytes`. You could use the same algorithm as you do now i.e. return record_bytes_ from lookahead_cache_ -> shift cache forward.